README.md
JeVe_EasyOTA library for Arduino
================================

# Purpose

Library which makes it easy to add support for Over-The-Air (OTA) updates to your project. Works with ESP8266 and ESP32.

## Features
- [x] mDNS
- [x] Add multiple networks to connect to
- [x] Allow/disallow open networks
- [x] Automatic reconnect on lost network
- [ ] HTTP update in AP mode
- [ ] Web update (requires web server)
- [ ] Stream update (e.g. MQTT or other)
- [ ] AdHoc network firmware sync - if you have ESP AdHoc network and update one device, all get updated
- [ ] Firmware sync - you can setup a bunch of devices on the same network you want to keep firmware synchronized without actual AdHoc

## How it works

This library tries to connect to configured APs first. Then if those fail ,tries to scan for open networks and connect to those. If it is not possible to connect to neither configured nor open networks - it falls back to AP mode.

# Installation

## Install as ZIP

* [Click here to download the ZIP file](https://github.com/jeroenvermeulen/JeVe_EasyOTA/archive/master.zip)
* Start the Arduino IDE
* Go to _Sketch > Include Library > Add .ZIP Library..._
* Select the ZIP file you just downloaded

## Install using Git (OSX / Linux)

* Open a terminal window
* Go to your Arduino libraries dir: `cd ~/Documents/Arduino/libraries`
* Execute: `git clone https://github.com/jeroenvermeulen/JeVe_EasyOTA.git`
* (Re)start your Arduino IDE

# Usage

The included examples are the easiest way to learn how to use this library. <br />
You will find them in the Arduino IDE under menu _File > Examples > Examples from Custom Libraries (bottom) > JeVe_EasyOTA_.

**WARNING:** Make sure you select the right board in the Arduino IDE because there is less checking when you upload using OTA.

After you put the code in your Arduino, after a few minutes you should see the OTA port in the Arduino IDE under the menu _Tools > Port_.

If somehow the wifi connection fails, the Arduino will become an access point with the configured hostname as SSID. In the examples the default hostname is `ota-flash-demo`.

![Arduino IDE Menu > Port](https://raw.githubusercontent.com/jeroenvermeulen/JeVe_EasyOTA/master/docs/menu_ota_port.png)

## To include it in your own project:

Put on top of your Arduino file:
```
#include <JeVe_EasyOTA.h>
EasyOTA OTA("arduino_hostname");
```

Put this in your `setup()` function. This example logs debugging output to Serial:
```
Serial.begin(9600);
// This callback will be called when EasyOTA has anything to tell you.
OTA.onMessage([](const String& message, int line) {
  Serial.println(message);
});
// Add networks you wish to connect to
OTA.addAP("wifi_ssid", "wifi_password");
// Allow open networks.
// NOTE: gives priority to configured networks
OTA.allowOpen(true);
// Set password for OTA programming
OTA.setPassword("OTApassword");
```

Put this in you `loop` function:
```
OTA.loop();
```

Or if you need timing:
```
static unsigned long last_m = millis();
unsigned long now = millis();
OTA.loop(now);

if (now - last_m > 1000) {
	last_m = now;
	// do something every second
}
```

That's all folks.

# Problem Solving

## Network port is not showing in IDE

* It sometimes takes a few minutes for the port to show in the Arduino IDE
* Check if the port is found, this also seems to speed up detection in the IDE:
  * In OSX on console: `dns-sd -B _arduino._tcp`
  * In Windows use [Bonjour browser](http://hobbyistsoftware.com/bonjourBrowser)
* Try to restart the Arduino board, wait 5 minutes, check if the _Network port_ shows up
* Try to restart your Arduino IDE, wait 5 minutes, check if the _Network port_ shows up
* On Windows 8.1 or older, mDNS does may not work, or only when you install [Bonjour Services](https://support.apple.com/kb/DL999?locale=en_US).
* Your may need op open 'mDNS' on your firewall: UDP port 5353

# Test Results

## Works with:

### ESP8266

#### Wemos® TTGO ESP8266 0.91 Inch OLED For Arduino Nodemcu

* Board selection: NodeMCU 0.9 (ESP-12 Module)
* Example: OTA_Oled_U8g2
* [Banggood board](https://www.banggood.com/Wemos-TTGO-ESP8266-0_91-Inch-OLED-For-Arduino-Nodemcu-p-1205904.html)

#### Geekcreit® NodeMcu Lua ESP8266 ESP-12E WIFI Development Board

* Board selection: NodeMCU 1.0 (ESP-12E Module)
* Example: OTA
* [Banggood board](https://www.banggood.com/Geekcreit-Doit-NodeMcu-Lua-ESP8266-ESP-12E-WIFI-Development-Board-p-985891.html)

#### Geekcreit® LoLin V3 NodeMcu Lua WIFI Development Board

* Board selection: NodeMCU 1.0 (ESP-12E Module)
* Example: OTA
* [Banggood board](https://www.banggood.com/V3-NodeMcu-Lua-WIFI-Development-Board-p-992733.html)

#### Upgraded Version 1M Flash ESP8266 ESP-01 WIFI Transceiver Wireless Module

* Important: Select _Tools > Flash Size > 1M (128K SPIFFS)_
* Board selection: Generic ESP8266
* Example: OTA
* [Banggood board](https://www.banggood.com/Upgraded-Version-1M-Flash-ESP8266-ESP-01-WIFI-Transceiver-Wireless-Module-p-979509.html)

#### WeMos® D1 mini V2.2.0 WIFI Internet Development Board Based ESP8266 4MB FLASH ESP-12S Chip

* Board selection: WeMos D1 R2 & Mini
* Example: OTA
* [Banggood board](https://www.banggood.com/WeMos-D1-mini-V2_2_0-WIFI-Internet-Development-Board-Based-ESP8266-4MB-FLASH-ESP-12S-Chip-p-1143874.html)

### ESP32

#### Geekcreit® ESP32 Development Board WiFi+Bluetooth Ultra Low Power Consumption Dual Cores ESP-32 ESP-32S Board

* Board selection: DOIT ESP32 DEVKIT V1
* Example: OTA
* [Banggood board](https://www.banggood.com/ESP32-Development-Board-WiFiBluetooth-Ultra-Low-Power-Consumption-Dual-Cores-ESP-32-ESP-32S-Board-p-1109512.html)

#### WeMos® TTGO ESP32 WiFi + Bluetooth Board 4MB Flash UNO D1 R32
##### with Geekcreit® IIC / I2C 1602 Blue Backlight LCD Display Module For Arduino

* Board selection: ESP32 Dev Module
* Example: OTA_LCD_I2C
* Note: The SCL and SCA pins are the ones closest to the reset button on this board.
* [Banggood board](https://www.banggood.com/ESP32-Development-Board-WiFiBluetooth-Ultra-Low-Power-Consumption-Dual-Cores-ESP-32-ESP-32S-Board-p-1109512.html)
* [Banggood display](https://www.banggood.com/IIC-I2C-1602-Blue-Backlight-LCD-Display-Module-For-Arduino-p-950726.html)


CHANGELOG.md
### Master

  * Remove some compiler warnings - thanks to @gordoste
  
### v3.1.0 (2018-09-27)
  
  * Optional progress bar by @gordoste

### v3.0.1 (2018-09-17)

  * Improved README.md
  * Tested on more boards
  * Removed extra comma in library.json - thanks @gordoste

### v3.0.0 (2018-02-19)

  * Refactored JeVe_EasyOTA Library
	* Updated examples
	* Updated documents
	* Added GPL license
	
### v3.0.0 (2018-02-19)

  * Refactored JeVe_EasyOTA Library
	* Updated examples
	* Updated documents
	* Added GPL license

### v2.2.0 (2017-11-21)

  * Improved README.md
  * Figured out how to get it working on the ESP-01 1M
  * Moved library source files to `src` dir
  * Added I2C LCD example
  * Shortened messages to fit on 16x2 LCD
  * Changed `architectures` value to `esp8266,esp32,espressif32`
  * Added library.json

### v2.1.1 (2017-11-20)

  * Added different hardware test results to README
  * Replaced `TelnetServer(8266)` by `ArduinoOTA.setPort(8266);`

### v2.1.0 (2017-11-20)

  * Added + tested support for ESP32
  * Improved display messages

### v2.0.1 (2017-11-20)

  * Added installation instructions to README.md
  * Added problem solving instructions

### v2.0.0 (2017-11-20)

  * Renamed from JV_OTA to JeVe_EasyOTA
  * Improved OTA_U8g2 example
  * Added README and CHANGELOG

### v1.0.0 (2017-11-18)

  * First working version
  * Added U8g2 example


JV_OTA.h
/*
 * JV_OTA.h - library to include to allow Over-The-Air updates of ESP8266
 *
 * Inspired on:
 *    http://simplestuffmatters.com/?p=69
 *    https://github.com/esp8266/Arduino/tree/master/libraries/ESP8266mDNS/examples/OTA-mDNS-SPIFFS
 *
 *  mDNS monitor (OSX):  dns-sd -B _arduino._tcp .
 *
 */

#ifndef JV_OTA_h
#define JV_OTA_h

#include <ESP8266WiFi.h>  // https://github.com/esp8266/Arduino/blob/master/libraries/ESP8266WiFi/src/ESP8266WiFi.h
#include <ESP8266mDNS.h>  // https://github.com/esp8266/Arduino/blob/master/libraries/ESP8266mDNS/ESP8266mDNS.h
#include <WiFiUdp.h>      // https://github.com/esp8266/Arduino/blob/master/libraries/ESP8266WiFi/src/WiFiUdp.h
#include <ArduinoOTA.h>   // https://github.com/esp8266/Arduino/blob/master/libraries/ArduinoOTA/ArduinoOTA.h

class JV_OTA
{
  public:
    typedef std::function<void(char *message, int line)> THandlerFunction_Message;

    JV_OTA(void);
    void setup(char* wifi_ssid, char* wifi_password = "", char* wifi_hostname = "");
    void loop();
    void onMessage(THandlerFunction_Message fn);

  private:
    char*  wifi_ssid;
    char*  wifi_password;
    char*  wifi_hostname;
    THandlerFunction_Message on_message;
    void showMessage(char *message, int line=1);
    void showMessage(String message, int line=1);
};

#endif

JV_OTA.cpp
/*
 * JV_OTA.cpp - library to include to allow Over-The-Air updates of ESP8266
 *
 * Inspired on:
 *    http://simplestuffmatters.com/?p=69
 *    https://github.com/esp8266/Arduino/tree/master/libraries/ESP8266mDNS/examples/OTA-mDNS-SPIFFS
 *
 *  mDNS monitor (OSX):  dns-sd -B _arduino._tcp .
 *
 */
 
#include "JV_OTA.h"

//Necesary to make Arduino Software autodetect OTA device
WiFiServer TelnetServer(8266);

JV_OTA::JV_OTA() {
  // Constructor
}

void JV_OTA::onMessage(THandlerFunction_Message fn) {
  on_message = fn;
}

void JV_OTA::setup(char* wifi_ssid, char* wifi_password, char* wifi_hostname) {
  this->wifi_ssid = wifi_ssid;
  this->wifi_password = wifi_password;
  this->wifi_hostname = wifi_hostname;
  showMessage("", 1); // New line in case of using serial output
  showMessage("Connecting Wifi:", 1);
  showMessage(this->wifi_ssid, 2);
  WiFi.hostname(this->wifi_hostname);
  WiFi.begin(this->wifi_ssid, this->wifi_password);
  unsigned long startTime = millis();
  String progressDots = "";
  while (WiFi.status() != WL_CONNECTED && millis() - startTime < 10000) {
    delay(500);
    progressDots += ".";
    showMessage(progressDots, 2);
  }

  if (WiFi.status() == WL_CONNECTED) {
    showMessage("IP Address:", 1);
    showMessage(WiFi.localIP().toString(), 2);
  } else {
    showMessage("Can't connect WiFi", 1);
    showMessage("Going into AP mode.", 2);
    WiFi.mode(WIFI_AP);
    delay(10);
    WiFi.softAP(this->wifi_hostname);
    showMessage("AP: " + String(this->wifi_hostname), 1);
    showMessage("IP: " + WiFi.softAPIP().toString(), 2);
  }

  TelnetServer.begin();  // Necesary to make Arduino Software autodetect OTA device

  ArduinoOTA.onStart([this]() {
    showMessage("OTA starting...", 1);
    showMessage("", 2);  // If display: Clean any previous error
  });
  ArduinoOTA.onEnd([this]() {
    showMessage("OTA finished.",1);
    showMessage("Rebooting...",2);
  });
  ArduinoOTA.onProgress([this](unsigned int progress, unsigned int total) {
    static unsigned int prevPerc = 100;
    unsigned int perc = (progress / (total / 100));
    unsigned int roundPerc = 5 * (int)(perc / 5);
    if ( roundPerc != prevPerc) {
      prevPerc = roundPerc;
      showMessage("OTA upload " + String(roundPerc) + "%");
    }
  });
  ArduinoOTA.onError([this](ota_error_t error) {
    showMessage("OTA Error " + String(error) + ":", 1);
    String line2 = "";
    if (error == OTA_AUTH_ERROR)         line2 = "Auth Failed";
    else if (error == OTA_BEGIN_ERROR)   line2 = "Begin Failed";
    else if (error == OTA_CONNECT_ERROR) line2 = "Connect Failed";
    else if (error == OTA_RECEIVE_ERROR) line2 = "Receive Failed";
    else if (error == OTA_END_ERROR)     line2 = "End Failed";
    showMessage(line2, 2);
  });
  ArduinoOTA.setHostname(this->wifi_hostname);
  ArduinoOTA.begin();
};

void JV_OTA::loop() {
  ArduinoOTA.handle();
}

void JV_OTA::showMessage(char *message, int line) {
  if (on_message) {
    on_message(message, line);
  }
}
void JV_OTA::showMessage(String message, int line) {
  showMessage((char *)message.c_str(), line);
}



src/EasyOTA.h
/*
 * EasyOTA.h - library to include to allow Over-The-Air updates of ESP8266 and ESP32
 *
 * Refactored from:
 *    https://github.com/jeroenvermeulen/JeVe_EasyOTA
 *
 *  Copyright (C) 2018  foxis (Andrius Mikonis <andrius.mikonis@gmail.com>)
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 *  mDNS monitor (OSX):  dns-sd -B _arduino._tcp
 *
 */

#ifndef EasyOTA_h
#define EasyOTA_h

#ifdef ESP8266
#include <ESP8266WiFi.h>  // https://github.com/esp8266/Arduino/blob/master/libraries/ESP8266WiFi/src/ESP8266WiFi.h
#include <ESP8266mDNS.h>  // https://github.com/esp8266/Arduino/blob/master/libraries/ESP8266mDNS/ESP8266mDNS.h
#endif

#ifdef ESP32
#include <WiFi.h>         // https://github.com/espressif/arduino-esp32/blob/master/libraries/WiFi/src/WiFi.h
#include <ESPmDNS.h>      // https://github.com/espressif/arduino-esp32/blob/master/libraries/ESPmDNS/src/ESPmDNS.h
#endif

#include <WiFiUdp.h>      // https://github.com/esp8266/Arduino/blob/master/libraries/ESP8266WiFi/src/WiFiUdp.h
#include <ArduinoOTA.h>   // https://github.com/esp8266/Arduino/blob/master/libraries/ArduinoOTA/ArduinoOTA.h

#include <map>

#define DEFAULT_OTA_PORT 8266
#define DEFAULT_OTA_HOSTNAME "EasyOta"

#define GETTER(T, name) T name() { return _##name; }
#define SETTER(T, name) T name(T name) { T pa##name = _##name; _##name = name; return pa##name; }

class EasyOTA
{
  public:
		typedef enum {
			EOS_UNKNOWN,
			EOS_SETUP_WIFI,
			EOS_SETUP_OTA,
			EOS_STA,
			EOS_AP,
			EOS_DISCONNECTED,
		} STATE;

		typedef std::function<void(const String& message, int line_nr)> THandlerFunction_Message;
		typedef std::function<void(const String& ssid, STATE state)> THandlerFunction_Connect;
		typedef std::function<void(const String& ssid, uint8_t sec, int32_t rssi, const uint8_t * BSSID, int32_t chan, bool hidden)> THandlerFunction_Scan;
    typedef std::function<void(const String& ssid, const String& passw, void * param)> THandlerFunction_APList;

		EasyOTA();
		EasyOTA(const String& hostname);
    EasyOTA(const String& hostname, int port);

		void addAP(const String& wifi_ssid, const String& wifi_password);
		void removeAP(const String& wifi_ssid);
		void eachAP(THandlerFunction_APList fn, void * param);
		void setPassword(const String& password);

		///
		/// one of these must be called inside main loop
		///
		void loop(unsigned long now);
    void loop();

		///
		/// Callback setters
		///
		void onMessage(THandlerFunction_Message fn);
		void onConnect(THandlerFunction_Connect fn);
    void onScan(THandlerFunction_Scan fn);

		///
		/// Allows open wifi networks
		/// NOTE: will scan for networks
		///
		GETTER(bool, allowOpen)
		SETTER(bool, allowOpen)

		///
		/// Specifies after how many tries to fallback to AP mode if no networks found
		///  maxRetries = 0 - go to AP mode if no networks available
		///  maxRetries = -1 - never go to AP mode
		///  maxRetries > 0 - number of attempts to find suitable network
		GETTER(int, maxRetries)
		SETTER(int, maxRetries)

		///
		/// Sets a timeout for connection to be established
		///
		GETTER(long, timeout)
		SETTER(long, timeout)

		///
		/// Sets an interval for scanning for networks while in AP mode
		///  scanInterval = 0  - always scan for networks while in AP mode
		///  scanInterval = -1 - once in AP mode, stay in AP mode
		///  scanInterval >0 - scan every so much milliseconds
		GETTER(long, scanInterval)
		SETTER(long, scanInterval)

		///
		/// Gets current Wifi STATE
		///
		GETTER(STATE, state)

		///
		/// Gets currently connected SSID or self hostname
		///
		GETTER(String, currentAP)

  private:
		STATE _state;
		bool _allowOpen;
		int _maxRetries;
		int _retries_current;
		bool _ap;
		long _timeout;
		long _scanInterval;

		std::map<String, String> _access_points;
		String _hostname;
		String _password;
		String _currentAP;
		int _port;

		THandlerFunction_Message on_message;
		THandlerFunction_Connect on_connect;
    THandlerFunction_Scan on_scan;

		void showMessage(const String& message, int line_nr=1);
    void callConnect(EasyOTA::STATE state);
		void onScan(const String& ssid, uint8_t sec, int32_t rssi, const uint8_t * BSSID, int32_t chan, bool hidden);

		int setupWifi(unsigned long now);
		int connectWifi(unsigned long now, const String& wifi_ssid, const String& wifi_password, const uint8_t * BSSID, int32_t chan);
		int scanWifi(unsigned long now);
		int setupOTA(unsigned long now);
};

#endif


src/JeVe_EasyOTA.cpp
/*
 * EasyOTA.cpp - library to include to allow Over-The-Air updates of ESP8266 and ESP32
 *
 * Inspired on:
 *    http://simplestuffmatters.com/?p=69
 *    https://github.com/esp8266/Arduino/tree/master/libraries/ESP8266mDNS/examples/OTA-mDNS-SPIFFS
 *    https://github.com/espressif/arduino-esp32/blob/master/libraries/ArduinoOTA/examples/BasicOTA/BasicOTA.ino
 *
 *  Copyright (C) 2018  Jeroen Vermeulen (info@jeve.eu), foxis (Andrius Mikonis <andrius.mikonis@gmail.com>)
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 *  mDNS monitor (OSX):  dns-sd -B _arduino._tcp
 *
 */

#include "JeVe_EasyOTA.h"
#include <set>

// Constructors
EasyOTA::EasyOTA() : EasyOTA(DEFAULT_OTA_HOSTNAME, DEFAULT_OTA_PORT)
{
}

EasyOTA::EasyOTA(const String& hostname) : EasyOTA(hostname, DEFAULT_OTA_PORT)
{
}

EasyOTA::EasyOTA(const String& hostname, int port) {
	_hostname = hostname;
	_port = port;
	_state = EOS_UNKNOWN;
	on_message = NULL;
	on_connect = NULL;
	on_scan = NULL;
	_timeout = 10000;
	_ap = false;
	_maxRetries = 0;
	_retries_current = 0;
	_scanInterval = -1;
}

void EasyOTA::onMessage(THandlerFunction_Message fn) {
  on_message = fn;
}
void EasyOTA::onConnect(THandlerFunction_Connect fn) {
  on_connect = fn;
}
void EasyOTA::onScan(THandlerFunction_Scan fn) {
  on_scan = fn;
}
void EasyOTA::setPassword(const String& password)
{
	_password = password;
}
void EasyOTA::addAP(const String& wifi_ssid, const String& wifi_password)
{
	_access_points.insert(std::pair<String, String>(wifi_ssid, wifi_password));
}
void EasyOTA::removeAP(const String& wifi_ssid)
{
	_access_points.erase(wifi_ssid);
}
void EasyOTA::eachAP(THandlerFunction_APList fn, void * param)
{
	if (fn == NULL)
		return;
	std::map<String, String>::iterator I = _access_points.begin();
	while (I != _access_points.end()) {
		fn(I->first, I->second, param);
		I++;
	}
}

int EasyOTA::setupOTA(unsigned long now)
{
	// ArduinoOTA callback functions
  ArduinoOTA.onStart([this]() {
    showMessage("OTA starting...", 1);
  });
  ArduinoOTA.onEnd([this]() {
    showMessage("OTA done.Reboot...",1);
  });
  #ifdef OTA_USE_PROGRESS_BAR
  static char progressBar[OTA_PROGRESS_BAR_SIZE+1];
  memset(progressBar, ' ', OTA_PROGRESS_BAR_SIZE);
  progressBar[0] = '[';
  progressBar[OTA_PROGRESS_BAR_SIZE-1] = ']';
  progressBar[OTA_PROGRESS_BAR_SIZE] = '\0';
  ArduinoOTA.onProgress([this](unsigned int progress, unsigned int total) {
	static unsigned int prevProg = 100;
    unsigned int prog = progress / (total/(OTA_PROGRESS_BAR_SIZE-2));
	Serial.println(prog);
	if ( prog != prevProg) {
      memset(progressBar+sizeof(char), '=', prog);
      showMessage(progressBar, 2);
    }
  });
  #else
  ArduinoOTA.onProgress([this](unsigned int progress, unsigned int total) {
    static unsigned int prevPerc = 100;
    unsigned int perc = (progress / (total / 100));
    unsigned int roundPerc = 5 * (int)(perc / 5);
    if ( roundPerc != prevPerc) {
      prevPerc = roundPerc;
      showMessage("OTA upload " + String(roundPerc) + "%", 1);
    }
  });
  #endif
  ArduinoOTA.onError([this](ota_error_t error) {
    showMessage("OTA Error " + String(error) + ":", 0);
    String line2 = "";
    if (error == OTA_AUTH_ERROR)         line2 = "Auth Failed";
    else if (error == OTA_BEGIN_ERROR)   line2 = "Begin Failed";
    else if (error == OTA_CONNECT_ERROR) line2 = "Connect Failed";
    else if (error == OTA_RECEIVE_ERROR) line2 = "Receive Failed";
    else if (error == OTA_END_ERROR)     line2 = "End Failed";
    showMessage(line2, 1);
  });

  ArduinoOTA.setPort(_port);
	if (_password != "")
		ArduinoOTA.setPassword(_password.c_str());
  ArduinoOTA.setHostname(_hostname.c_str());
  ArduinoOTA.begin();

	return 0;
}

int EasyOTA::connectWifi(unsigned long startTime, const String& wifi_ssid, const String& wifi_password, const uint8_t * BSSID, int32_t chan)
{
	wl_status_t status;

	showMessage("Trying " + wifi_ssid, 0);

	if (BSSID)
		WiFi.begin(wifi_ssid.c_str(), wifi_password.c_str(), chan, BSSID);
	else
		WiFi.begin(wifi_ssid.c_str(), wifi_password.c_str());
	do {
		status = WiFi.status();

		switch (status){
			case WL_CONNECTED:
				showMessage("IP " + WiFi.localIP().toString(), 0);
				showMessage("SSID " + wifi_ssid, 1);
				_ap = false;
				_currentAP = wifi_ssid;
				return 0;
			case WL_NO_SSID_AVAIL:
			case WL_CONNECT_FAILED:
				break;
			default:
				break;
		}
		delay(10);
	} while (millis() < startTime + _timeout);

	return 1;
}

int EasyOTA::setupWifi(unsigned long now) {
	static unsigned long retry_ms;

	if (_access_points.size() == 1 && !_allowOpen) {
		// Skip scanning phase for single ap
		std::map<String, String>::iterator I = _access_points.begin();
		if (!connectWifi(now, I->first, I->second, NULL, 0))
			return 0;
	} else if (_access_points.size() > 1 || _allowOpen) {
		if (WiFi.status() == WL_CONNECTED)
		{
			WiFi.mode(WIFI_STA);
			WiFi.disconnect();
		}

		int result = scanWifi(now);
		switch (result) {
			case 0: return 0; // connected
			case 1:
				WiFi.scanDelete();
				return 1; // scanning/connecting
			default:
				WiFi.scanDelete();
				break; // no wifi available, fallback to AP mode
		}
	}

	if (_retries_current == 0)
		retry_ms = now;

	if (now <= retry_ms + 10 * _timeout)
		return 1;

	if (_maxRetries < 0)
	{
		retry_ms = now;
		return 1;
	}

	if (_retries_current < _maxRetries) {
		_retries_current++;
		retry_ms = now;
		return 1;
	}

	// fallback to AP mode
	showMessage("No AP available", 0);
	showMessage("Starting AP", 1);
	WiFi.mode(WIFI_AP);
	WiFi.softAP(_hostname.c_str());
	showMessage("AP " + _hostname, 0);
	showMessage("IP " + WiFi.softAPIP().toString(), 1);
	_ap = true;
	_currentAP = _hostname;
	return 0;
};

int EasyOTA::scanWifi(unsigned long now)
{
	wl_status_t status = WiFi.status();

	if (status != WL_CONNECTED)
	{
		int8_t scanResult = WiFi.scanComplete();
		if (scanResult == WIFI_SCAN_RUNNING) {
			return 1;
		} else {
			if (scanResult <= 0) {
				showMessage("Wifi scan", 0);
				WiFi.scanNetworks(true);
				return 1;
			} else {
				String bestSSID;
				int32_t bestRSSI = -1000;
				uint8 bestBSSID[6];
				int32_t bestChannel;
				std::set<String> black_list;
				bool call_scan = true;

				while (true) {
					for(int8_t i = 0; i < scanResult; ++i) {
						String ssid_scan;
						int32_t rssi_scan;
						uint8_t sec_scan;
						uint8_t* BSSID_scan;
						int32_t chan_scan;
						bool hidden_scan;

						WiFi.getNetworkInfo(i, ssid_scan, sec_scan, rssi_scan, BSSID_scan, chan_scan, hidden_scan);
						if (call_scan)
							onScan(ssid_scan, sec_scan, rssi_scan, BSSID_scan, chan_scan, hidden_scan);

						bool bl = black_list.count(ssid_scan);
						bool cap = _access_points.count(ssid_scan);
						bool exhausted = true;
						// prioritize configured ap
						std::map<String, String>::iterator I = _access_points.begin();
						while (I != _access_points.end()) {
							exhausted = black_list.count(I->first);
							if (!exhausted)
								break;
							I++;
						}

						// skip black-listed networks (those that failed to connect to)
						if (rssi_scan > bestRSSI && !bl) {
							if ((sec_scan == ENC_TYPE_NONE && _allowOpen && exhausted) || cap) {
								bestSSID = ssid_scan;
								bestRSSI = rssi_scan;
								bestChannel = chan_scan;
								memcpy(bestBSSID, BSSID_scan, sizeof(bestBSSID));
							}
						}
					}
					call_scan = false;

					// some network found
					if (bestRSSI != -1000) {
						std::map<String, String>::iterator f = _access_points.find(bestSSID);
						int result = connectWifi(now, bestSSID, f != _access_points.end() ? _access_points[bestSSID] : "", bestBSSID, bestChannel);
						if (result)
						{
							black_list.insert(bestSSID);
							bestRSSI = -1000;
						}
						else
						{
							return 0;
						}
					}
					else
						return 2; // no network found. fallback to AP mode
				}
			}
		}
	}
	// showMessage("should not go here", 1);
	return 0;
}

void EasyOTA::loop() {
	loop(millis());
}

void EasyOTA::loop(unsigned long now) {
	unsigned long rescan_ms = 0;

	switch (_state)
	{
		case EOS_UNKNOWN:
			WiFi.mode(WIFI_STA);
#ifdef ESP8266
		  WiFi.hostname(_hostname);
#elif ESP32
		  WiFi.setHostname(_hostname);
#endif
			_retries_current = 0;
			_ap = false;
			_currentAP = "";
			_state = EOS_SETUP_WIFI;
			break;
		case EOS_SETUP_WIFI:
			if (!setupWifi(now))
				_state = EOS_SETUP_OTA;
			break;
		case EOS_SETUP_OTA:
			if (!setupOTA(now))
				callConnect(_ap ? EOS_AP : EOS_STA);
			rescan_ms = now;
			break;
		case EOS_DISCONNECTED:
			// TODO Prepare for reconnect
			_state = EOS_UNKNOWN;
			break;
		default:
			if (WiFi.status() == WL_CONNECTED || _ap)
				ArduinoOTA.handle();
			else
			{
				callConnect(EOS_DISCONNECTED);
			}

			if (_ap && now > rescan_ms + _scanInterval && _scanInterval > 0)
			{
				// TODO: scan for available wifi and connect to it
				rescan_ms = now;
			}
	}
}

void EasyOTA::showMessage(const String& message, int line_nr) {
  if (on_message) {
    on_message(message, line_nr);
  }
}

void EasyOTA::callConnect(STATE state) {
	STATE last_state = _state;
	_state = state;

	if (on_connect && state != last_state)
	{
		on_connect(_currentAP, state);
	}
}
void EasyOTA::onScan(const String& ssid, uint8_t sec, int32_t rssi, const uint8_t * BSSID, int32_t chan, bool hidden)
{
	if (on_scan)
	{
		on_scan(ssid, sec, rssi, BSSID, chan, hidden);
	}
}
